pylot/utils.py

```diff
@@ -64,6 +64,170 @@ def __str__(self):
+        wx, wy, wz = w * x2, w * y2, w * z2
+        m = np.array([[1.0 - (yy + zz), xy - wz, xz + wy],
+                      [xy + wz, 1.0 - (xx + zz), yz - wx],
+                      [xz - wy, yz + wx, 1.0 - (xx + yy)]])
+        return m
+
+    @classmethod
+    def from_rotation(cls, rotation):
+        """Creates a Quaternion from a rotation including pitch, roll, yaw.
+
+        Args:
+            rotation (:py:class:`.Rotation`): A pylot rotation representing
+                the rotation of the object in degrees.
+
+        Returns:
+            :py:class:`.Quaternion`: The quaternion representation of the
+            rotation.
+        """
+        roll_by_2 = np.radians(rotation.roll) / 2.0
+        pitch_by_2 = np.radians(rotation.pitch) / 2.0
+        yaw_by_2 = np.radians(rotation.yaw) / 2.0
+
+        cr, sr = np.cos(roll_by_2), np.sin(roll_by_2)
+        cp, sp = np.cos(pitch_by_2), np.sin(pitch_by_2)
+        cy, sy = np.cos(yaw_by_2), np.sin(yaw_by_2)
+
+        w = cr * cp * cy + sr * sp * sy
+        x = cr * sp * sy - sr * cp * cy
+        y = -cr * sp * cy - sr * cp * sy
+        z = cr * cp * sy - sr * sp * cy
+
+        return cls(w, x, y, z)
+
+    @classmethod
+    def from_angular_velocity(cls, angular_velocity, dt):
+        """Creates a Quaternion from an angular velocity vector and the time
+        delta to apply it for.
+
+        Args:
+            angular_velocity (:py:class:`.Vector3D`): The vector representing
+                the angular velocity of the object in the body-frame.
+            dt (float): The time delta to apply the angular velocity for.
+
+        Returns:
+            :py:class:`.Quaternion`: The quaternion representing the rotation
+                undergone by the object with the given angular velocity in the
+                given delta time.
+        """
+        angular_velocity_np = angular_velocity.as_numpy_array() * dt
+        magnitude = np.linalg.norm(angular_velocity_np)
+
+        w = np.cos(magnitude / 2.0)
+        if magnitude < 1e-50:
+            # To avoid instabilities and nan.
+            x, y, z = 0, 0, 0
+        else:
+            imaginary = angular_velocity_np / magnitude * np.sin(
+                magnitude / 2.0)
+            x, y, z = imaginary
+        return cls(w, x, y, z)
+
+    def as_rotation(self):
+        """Retrieve the Quaternion as a Rotation in degrees.
+
+        Returns:
+            :py:class:`.Rotation`: The euler-angle equivalent of the
+                Quaternion in degrees.
+        """
+        SINGULARITY_THRESHOLD = 0.4999995
+        RAD_TO_DEG = (180.0) / np.pi
+
+        singularity_test = self.z * self.x - self.w * self.y
+        yaw_y = 2.0 * (self.w * self.z + self.x * self.y)
+        yaw_x = (1.0 - 2.0 * (self.y**2 + self.z**2))
+
+        pitch, yaw, roll = None, None, None
+        if singularity_test < -SINGULARITY_THRESHOLD:
+            pitch = -90.0
+            yaw = np.arctan2(yaw_y, yaw_x) * RAD_TO_DEG
```

Comment:
Why not use np.degrees?

Reply:
Because I just copied this version from UnrealEngine and didn't want to mess with anything :)